{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot the pangenome results\n",
    "- plot the taxonomic boxplots for pigs vs human\n",
    "- accumulation curve for pig v human\n",
    "\n",
    "Should be able to make taxonomy boxplots for each taxonomic rank"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ran the pig fastgather --> tax on all 338 SRAs. Subset to get 100 v 100\n",
    "# pig df, was done on all 399 datasets\n",
    "# Subset to 100 and save as file for input to compare to human\n",
    "dfp = pd.read_csv('../results/sourmash//fastgather/tax/pig_339.tax.csv')\n",
    "# fg does something weird with scaled, when sigs are not same scale as --scaled parameter\n",
    "dfp['f_weighted_at_rank']= dfp['f_weighted_at_rank']*10\n",
    "query_100 = dfp['query_name'].unique()\n",
    "# subset 100 random SRAs and to df. Use randomized df\n",
    "query_100_select = pd.Series(query_100).sample(100, random_state=42).tolist()\n",
    "dfp = dfp[dfp['query_name'].isin(query_100_select)]\n",
    "dfp.to_csv('./pig_100.tax.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the dataframes, and clean them \n",
    "dfh = pd.read_csv('./human_100.tax.csv')\n",
    "dfp = pd.read_csv('./pig_100.tax.csv')\n",
    "\n",
    "# define the taxonomic rank for all functions later\n",
    "taxrank = 'genus'\n",
    "\n",
    "# remove unclassified, keep only columns of interest, only keep certain taxonomic rank\n",
    "dfh = remove_cols(dfh, taxrank)\n",
    "dfp = remove_cols(dfp, taxrank)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ACCUMULATION CURVE\n",
    "#plot accumulationcurve for taxonomic rank defined above\n",
    "plot_accum2(dfp, dfh, 'accum.png', 'pig', 'human')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Plot abundance weighted fractions for taxonomic rank in human vs pig (in boxplots)\n",
    "# col for source, so we can plot later\n",
    "dfp['source'] = 'pig'\n",
    "dfh['source'] = 'human'\n",
    "\n",
    "# what genera are most abundant for either \n",
    "dfp = split_lineage(dfp, taxrank)\n",
    "dfh = split_lineage(dfh, taxrank)\n",
    "\n",
    "# concat df, so have tax for both pigs and humans, with columns that describes what metaG type the species was found in\n",
    "df = pd.concat([dfh, dfp], ignore_index=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep only tax lineages that are in > 95 of the metaGs\n",
    "pig_ab = calc_num_metag(dfp, taxrank, 95)\n",
    "human_ab = calc_num_metag(dfh, taxrank, 95)\n",
    "# combine the top for human and pig\n",
    "top_genera_combined = set(pig_ab) | set(human_ab)\n",
    "\n",
    "# Only keep the tax lineages that are in this top\n",
    "dfp = dfp[dfp[taxrank].isin(top_genera_combined)]\n",
    "dfh = dfh[dfh[taxrank].isin(top_genera_combined)]\n",
    "df = pd.concat([dfh, dfp], ignore_index=True)\n",
    "df = df[['query_name', taxrank, 'source', 'f_weighted_at_rank']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## BOXPLOT\n",
    "# plot the f_weigthed for pigs v humans in a boxplot\n",
    "compare_boxplot(df, taxrank, 'genus_pig.v.human.png')\n",
    "\n",
    "# or boxplot just pigs or just human\n",
    "boxplotmetagperc_1df(dfh, taxrank, 'genus_human.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_cols(df, rank):\n",
    "    \"\"\" Creates dataframe with only taxonomic info of interest. Only keeps gather columns of interest \"\"\"\n",
    "    df = df[df['lineage'] != 'unclassified']\n",
    "    df = df[['query_name', 'rank', 'lineage', 'f_weighted_at_rank']]\n",
    "    df['lineage'] = df['lineage'].str.strip()\n",
    "    # for now keep speces level matches\n",
    "    df = df[df['rank'] == rank]\n",
    "    return df\n",
    "\n",
    "# plot accumulation curve for all species recovered in all metaGs\n",
    "def plot_accum(df, outplot):\n",
    "    \"\"\" Plot one accumulation curve, at specified taxonomic rank. \"\"\"\n",
    "    df = df.lineage.value_counts().to_frame().reset_index()\n",
    "    df.rename(columns={'count':'num_metaG'}, inplace=True)\n",
    "    df = df[df['num_metaG'] > 0]\n",
    "    sns.ecdfplot(data=df, x=\"num_metaG\", stat='percent').set(xlabel = \"num_metaGs\", ylabel= \"% species recovered\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outplot)\n",
    "\n",
    "# plot both curves, add subject names (human/chicken/pig/cow whatever)\n",
    "def plot_accum2(df1, df2, outplot, subject1, subject2):\n",
    "    \"\"\" Plot two accumulation curves, for metagenomes from differents species \"\"\"\n",
    "    df1 = df1.lineage.value_counts().to_frame().reset_index()\n",
    "    df2 = df2.lineage.value_counts().to_frame().reset_index()\n",
    "    df1.rename(columns={'count':'num_metaG'}, inplace=True)\n",
    "    df2.rename(columns={'count':'num_metaG'}, inplace=True)\n",
    "    df1 = df1[df1['num_metaG'] > 0]\n",
    "    df2 = df2[df2['num_metaG'] > 0]\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.ecdfplot(data=df1, x=\"num_metaG\", stat='percent', label=subject1)\n",
    "    sns.ecdfplot(data=df2, x=\"num_metaG\", stat='percent', label=subject2, color='orange')  \n",
    "    plt.xlabel(\"num_metaGs\")\n",
    "    plt.ylabel(\"% species recovered\")\n",
    "    plt.legend()\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outplot)\n",
    "    plt.show()\n",
    "\n",
    "def split_lineage(df, rank):\n",
    "    \"\"\" Creates df with for each taxonomic rank the specific columns \"\"\"\n",
    "    split_df = df['lineage'].str.split(';', expand=True)\n",
    "    # Create the appropriate columns based on the rank\n",
    "    if rank == 'species':\n",
    "        df[['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species']] = split_df\n",
    "    elif rank == 'genus':\n",
    "        df[['domain', 'phylum', 'class', 'order', 'family', 'genus']] = split_df\n",
    "    elif rank == 'family':\n",
    "        df[['domain', 'phylum', 'class', 'order', 'family']] = split_df\n",
    "    else:\n",
    "        raise ValueError(\"Rank must be 'genus' or 'family'.\")\n",
    "    \n",
    "    return df\n",
    "\n",
    "\n",
    "def calc_num_metag(df, rank, n):\n",
    "\"\"\" Only keeps taxonomic lineages that were found in more than n metagenomes\"\"\"\n",
    "    counts = df[rank].value_counts()\n",
    "    valid_strings = counts[counts > n].index\n",
    "    return valid_strings\n",
    "\n",
    "# boxplot based off highest abundance\n",
    "def compare_boxplot(df, rank, outplot):\n",
    "    means = df.groupby(rank)['f_weighted_at_rank'].mean().sort_values()\n",
    "    ordered_rank = means.index\n",
    "    counts = df.groupby([rank, 'source']).size().reset_index(name='count')\n",
    "    plt.figure(figsize=(12, 8))\n",
    "    sns.set_style(\"whitegrid\")\n",
    "    ax = sns.boxplot(x=rank, y='f_weighted_at_rank', data=df, order=ordered_rank, hue='source', palette='Set1', showfliers=False)\n",
    "\n",
    "    # Add text annotations for counts\n",
    "    for i, phylo_rank in enumerate(ordered_rank):\n",
    "        for j, source in enumerate(df['source'].unique()):\n",
    "            count = counts[(counts[rank] == phylo_rank) & (counts['source'] == source)]['count'].values\n",
    "            if len(count) > 0:\n",
    "                count = count[0]\n",
    "                y_pos = df[(df[rank] == phylo_rank) & (df['source'] == source)]['f_weighted_at_rank'].quantile(0.75)\n",
    "                # Calculate the x position based on the hue\n",
    "                x_pos = i + (j - 0.5) * 0.2  # Adjust the 0.2 value if needed for better spacing\n",
    "                ax.text(x_pos, y_pos + 0.05, f'n: {count}', horizontalalignment='center', size='medium', color='black', rotation=90)\n",
    "\n",
    "    ax.set_xticklabels(ax.get_xticklabels(), rotation=90)\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outplot)\n",
    "    plt.show()\n",
    "\n",
    "# boxplots per rank, how much of the metaG does rank make up?\n",
    "def boxplotmetagperc_1df(df, rank, outplot):\n",
    "    counts = df[rank].value_counts()\n",
    "    valid_strings = counts[counts > 90].index\n",
    "    filtered_df = df[df[rank].isin(valid_strings)]\n",
    "    medians = filtered_df.groupby(rank)['f_weighted_at_rank'].mean().sort_values()\n",
    "    ordered_categories = medians.index\n",
    "    counts = filtered_df[rank].value_counts().sort_index()\n",
    "\n",
    "    means = filtered_df.groupby(rank)['f_weighted_at_rank'].mean().sort_values()\n",
    "    nobs = filtered_df[rank].value_counts().reindex(means.index).values\n",
    "    nobs = [\"n: \" + str(x) for x in nobs]\n",
    "    plt.figure(figsize=(8, 8))  # Adjust size if necessary\n",
    "    sns.set_style(\"whitegrid\") \n",
    "    ax = sns.boxplot(x=rank, y='f_weighted_at_rank', data=filtered_df, showfliers=False,order=ordered_categories)\n",
    "    ax.set_xticklabels(ax.get_xticklabels(),rotation=90)\n",
    "    # Add it to the plot\n",
    "    pos = range(len(nobs))\n",
    "    for tick,label in zip(pos,ax.get_xticklabels()):\n",
    "        ax.text(pos[tick],\n",
    "                means[tick] + 0.1,\n",
    "                nobs[tick],\n",
    "                horizontalalignment='center',\n",
    "                rotation=45,\n",
    "                size='small',\n",
    "                color='black',\n",
    "                weight='semibold')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(outplot)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
