{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Microbial pangenomics\n",
    "- want to compare the results from viral pang to bacterial\n",
    "- take 5 species, that are common in most metaGs\n",
    "- cross them with the 10 test metaGs\n",
    "\n",
    "\n",
    "Lactobacillus amylovorus, Limosilactobacillus reuteri, Escherichia coli, RUG472 sp900545265, Bariatricus sp004560705\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# link dbs\n",
    "ln -s /group/ctbrowngrp/sourmash-db/gtdb-rs214/gtdb-rs214-k21.zip .\n",
    "ln -s /group/ctbrowngrp/sourmash-db/gtdb-rs214/gtdb-rs214.lineages.csv ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    " \n",
    "sourmash tax grep -i 'Blautia_A sp003471165' \\\n",
    "-t gtdb-rs214.lineages.csv -r species -o s__ecoli.csv\n",
    "sourmash tax grep -i 'lactobacillus amylovorus' \\\n",
    "-t gtdb-rs214.lineages.csv -r species -o s__lamylovorus.csv\n",
    "sourmash tax grep -i 'limosilactobacillus reuteri' \\\n",
    "-t gtdb-rs214.lineages.csv -r species -o s__lreuteri.csv\n",
    "sourmash tax grep -i 'RUG472 sp900545265' \\\n",
    "-t gtdb-rs214.lineages.csv -r species -o s__RUG472.csv\n",
    "sourmash tax grep -i 'Bariatricus sp004560705' \\\n",
    "-t gtdb-rs214.lineages.csv -r species -o s__bariatricus.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sourmash sig cat --picklist s__ecoli.csv:ident:ident \\\n",
    "gtdb-rs214-k21.zip -o gtdb.ecoli.k21.sig.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Snakefile\n",
    "import os\n",
    "import pandas as pd\n",
    "\n",
    "# Define samples\n",
    "CLUSTERS, = glob_wildcards('../results/microbial_pangenomic_test/s__{ident}.csv')\n",
    "METAGS = ['SRR8960980', 'SRR8960440', 'SRR8960721', 'SRR8960326', 'SRR8960915', 'SRR8960631', 'SRR11125655', 'SRR8960731', 'SRR8960200', 'SRR8960303']\n",
    "\n",
    "GTDB = '/group/ctbrowngrp/sourmash-db/gtdb-rs214/gtdb-rs214-k21.zip'\n",
    "\n",
    "\n",
    "wildcard_constraints:\n",
    "    sample='\\w+',\n",
    "\n",
    "rule all:\n",
    "    input:\n",
    "        expand('../results/microbial_pangenomic_test/classify/{ident}x{metag}.txt', ident=CLUSTERS, metag=METAGS),\n",
    "        expand('../results/microbial_pangenomic_test/classify/{ident}.pang.txt', ident=CLUSTERS),\n",
    "\n",
    "\n",
    "# filter contigs per cluster\n",
    "rule pick_contigs:\n",
    "    input: \n",
    "        picklist = \"../results/microbial_pangenomic_test/s__{ident}.csv\"\n",
    "    output: \n",
    "        sig = \"../results/microbial_pangenomic_test/{ident}.zip\"\n",
    "    conda: \n",
    "        \"branchwater\"\n",
    "    shell:\n",
    "        \"\"\"\n",
    "        sourmash sig cat --picklist {input.picklist}:ident:ident \\\n",
    "        {GTDB} -o {output.sig} \n",
    "        \"\"\"\n",
    "\n",
    "# merge sketches into pangenome\n",
    "rule pang_merge:\n",
    "    input:\n",
    "       sig='../results/microbial_pangenomic_test/{ident}.zip',\n",
    "    output:\n",
    "        pang = \"../results/microbial_pangenomic_test/{ident}.pang.sig.gz\",\n",
    "    conda: \n",
    "        \"branchwater\"\n",
    "    threads: 1\n",
    "    shell:\n",
    "        \"\"\" \n",
    "        sourmash scripts pangenome_merge \\\n",
    "        {input.sig} -o {output.pang} -k 21 \n",
    "        \"\"\"\n",
    "\n",
    "# create a ranktable\n",
    "rule ranktable:\n",
    "    input:\n",
    "       sig='../results/microbial_pangenomic_test/{ident}.pang.sig.gz',\n",
    "    output:\n",
    "        rankt = \"../results/microbial_pangenomic_test/{ident}.rankt.csv\",\n",
    "    conda: \n",
    "        \"branchwater\"\n",
    "    threads: 1\n",
    "    shell:\n",
    "        \"\"\" \n",
    "        sourmash scripts pangenome_ranktable {input.sig} \\\n",
    "        -o {output.rankt} -k 21\n",
    "        \"\"\"\n",
    "\n",
    "# Now do the pangenome classify for 10 metagenomes\n",
    "rule classify:\n",
    "    input:\n",
    "       rankt = \"../results/microbial_pangenomic_test/{ident}.rankt.csv\",\n",
    "       metag = \"../results/sourmash/sketches/read_s100/{metag}.sig.gz\" \n",
    "    output:\n",
    "        txt = \"../results/microbial_pangenomic_test/classify/{ident}x{metag}.txt\",\n",
    "    conda: \n",
    "        \"branchwater\"\n",
    "    threads: 1\n",
    "    shell:\n",
    "        \"\"\" \n",
    "        sourmash scripts pangenome_classify \\\n",
    "        {input.metag} {input.rankt} -k 21  > {output.txt}\n",
    "        \"\"\"\n",
    "rule classify_pang:\n",
    "    input:\n",
    "       rankt = \"../results/microbial_pangenomic_test/{ident}.rankt.csv\",\n",
    "       sig = \"../results/microbial_pangenomic_test/{ident}.pang.sig.gz\" \n",
    "    output:\n",
    "        txt = \"../results/microbial_pangenomic_test/classify/{ident}.pang.txt\",\n",
    "    conda: \n",
    "        \"branchwater\"\n",
    "    threads: 1\n",
    "    shell:\n",
    "        \"\"\" \n",
    "        sourmash scripts pangenome_classify \\\n",
    "        {input.sig} {input.rankt} -k 21 > {output.txt}\n",
    "        \"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add new MAGs from pig so we can see the pangenome sketches from those\n",
    "# 7 matches\n",
    "sourmash sig extract \\\n",
    " --picklist F_prausnitzii_new.txt:user_genome:name \\\n",
    " MAGs.drep999.zip -o F_prausnitzii_new.zip\n",
    "\n",
    "# 43 matches\n",
    "sourmash sig extract \\\n",
    " --picklist E_coli_new.txt:user_genome:name \\\n",
    " MAGs.drep999.zip -o E_coli_new.zip\n",
    "\n",
    "sourmash sig cat Faecalibacterium_prausnitzii_E.zip F_prausnitzii_new.zip -o F_prausnitzii.all.zip\n",
    "sourmash sig cat Escherichia_coli.zip E_coli_new.zip -o E_coli_.all.zip"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sourmash scripts pangenome_ranktable E_coli_new.zip -o E_coli_new.rankt.csv -k 21\n",
    "sourmash scripts pangenome_ranktable F_prausnitzii_new.zip -o F_prausnitzii.new.rankt.csv -k 21\n",
    "\n",
    "\n",
    "python ../../../workflow/scripts/calc-hash-presence.py E_coli_new.rankt.csv \\\n",
    "../../../workflow/input_files_pig_100.txt --scaled=1000 -k 21 -o E_coli_new.x.pig.dump\n",
    "\n",
    "python ../../../workflow/scripts/calc-hash-presence.py E_coli_new.rankt.csv \\\n",
    "../../../workflow/input_files_human.txt --scaled=1000 -k 21 -o E_coli_new.x.human.dump\n",
    "\n",
    "python ../../../workflow/scripts/calc-hash-presence.py \\\n",
    "F_prausnitzii.rankt.csv ../../../workflow/input_files_pig_100.txt \\\n",
    "--scaled=1000 -k 21 -o F_prausnitzii.x.pig.dump\n",
    "\n",
    "\n",
    "python ../../../workflow/scripts/parse-dump.py \\\n",
    "--dump-files-1 E_coli_new.x.human.dump \\\n",
    "--dump-files-2 E_coli_new.x.pig.dump > E_coli_new.cmp.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "python scripts/calc-hash-presence.py ../results/pangenome/pangenome_sketch/Lactobacillus_amylovorus.rankt.csv \\\n",
    "input_files_pig_100.txt --scaled=1000 -k 21 \\\n",
    "-o ../results/pangenome/dmp_pig/Lactobacillus_amylovorus.x.pig.dump\n",
    "\n",
    "python ../../workflow/scripts/parse-dump.py \\\n",
    "--dump-files-1 dmp_human/Lactobacillus_amylovorus.x.human.dump \\\n",
    "--dump-files-2 dmp_pig/Lactobacillus_amylovorus.x.pig.dump > dmp/Lactobacillus_amylovorus.cmp.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# one more for all the ecoli genomes\n",
    "sourmash scripts pangenome_ranktable E_coli.all.zip -o E_coli.all.rankt.csv -k 21\n",
    "\n",
    "python ../../../workflow/scripts/calc-hash-presence.py E_coli.all.rankt.csv \\\n",
    "../../../workflow/input_files_pig_100.txt --scaled=1000 -k 21 -o E_coli_all.x.pig.dump && \\\n",
    "python ../../../workflow/scripts/calc-hash-presence.py E_coli.all.rankt.csv \\\n",
    "../../../workflow/input_files_human.txt --scaled=1000 -k 21 -o E_coli_all.x.human.dump && \\\n",
    "python ../../../workflow/scripts/parse-dump.py \\\n",
    "--dump-files-1 E_coli_all.x.human.dump \\\n",
    "--dump-files-2 E_coli_all.x.pig.dump > E_coli_all.cmp.tsv"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
